File: programacion1/ejercicio7/main.c

/* Realizar un programa que dado un numero N, introducido como dato, escriba todos los numeros 
comprendidos entre 1 y 10000 que cumplan las dos reglas siguiente:
1. regla: La suma de sus cifras debe ser un divisor de N
2. regla: El producto de sus cifras debe ser un multiplo de N */

#include <stdio.h>
#include <stdlib.h>
#include "funciones.h"
#define MAX_VECTOR_RES 1000

int main(int argc, char const *argv[])
{
    int N = atoi(argv[1]);
    int j = 0;
    int buffer_res[MAX_VECTOR_RES]; // vector para guardar los resultados
    // inicializamos a 0 el vector de resultados
    for (int i = 0; i < MAX_VECTOR_RES; i++)
    {
        buffer_res[i] = 0;
    }

    for (int i = 1; i < 10000; i++)
    {
        if (result(N, i))
        {
            buffer_res[j] = i;
            j++;
        }
    }

    // mostramos el resultado en pantalla
    for (int i = 0; i < MAX_VECTOR_RES; i++)
    {
        printf("%d", buffer_res[i]);
    }
    printf("\n");
}

File: programacion1/ejercicio7/main.py
""" Realizar un programa que dado un numero N, introducido como dato, escriba todos los numeros 
comprendidos entre 1 y 10000 que cumplan las dos reglas siguiente:
1. regla: La suma de sus cifras debe ser un divisor de N
2. regla: El producto de sus cifras debe ser un multiplo de N """

from sys import argv

# devuelve la suma de los digitos de un numero n
def addNumber (n:int)->int:
    num_string = str(n)
    sum = 0
    for digit in num_string:
       sum += int(digit)
    return sum

# devuelve el producto de los digitos de un numero n

def mulNumber(n:int)->int:
    num_string = str(n)
    mul = 1
    for digit in num_string:
        mul *= int(digit)
    return mul

if __name__ == '__main__':
    
    N = int(argv[1])
    for i in range(1,10000):
        if (N % (addNumber(i))==0) and ((mulNumber(i) % N) ==0):
            print(f"{i}")

    
    
